src/utils/multipart-parser.js

```diff
@@ -0,0 +1,440 @@
+					mark('partData');
+					// fallsthrough
+				case S.PART_DATA:
+					previousIndex = index;
+
+					if (index === 0) {
+						// boyer-moore derrived algorithm to safely skip non-boundary data
+						i += boundaryEnd;
+						while (i < bufferLength && !(ui8a[i] in boundaryChars)) {
+							i += boundaryLength;
+						}
+
+						i -= boundaryEnd;
+						c = ui8a[i];
+					}
+
+					if (index < boundary.length) {
+						if (boundary[index] === c) {
+							if (index === 0) {
+								dataCallback('partData', true);
+							}
+
+							index++;
+						} else {
+							index = 0;
+						}
+					} else if (index === boundary.length) {
+						index++;
+						if (c === CR) {
+							// CR = part boundary
+							flags |= F.PART_BOUNDARY;
+						} else if (c === HYPHEN) {
+							// HYPHEN = end boundary
+							flags |= F.LAST_BOUNDARY;
+						} else {
+							index = 0;
+						}
+					} else if (index - 1 === boundary.length) {
+						if (flags & F.PART_BOUNDARY) {
+							index = 0;
+							if (c === LF) {
+								// unset the PART_BOUNDARY flag
+								flags &= ~F.PART_BOUNDARY;
+								callback('partEnd');
+								callback('partBegin');
+								state = S.HEADER_FIELD_START;
+								break;
+							}
+						} else if (flags & F.LAST_BOUNDARY) {
+							if (c === HYPHEN) {
+								callback('partEnd');
+								callback('end');
+								state = S.END;
+								flags = 0;
+							} else {
+								index = 0;
+							}
+						} else {
+							index = 0;
+						}
+					}
+
+					if (index > 0) {
+						// when matching a possible boundary, keep a lookbehind reference
+						// in case it turns out to be a false lead
+						lookbehind[index - 1] = c;
+					} else if (previousIndex > 0) {
+						// if our boundary turned out to be rubbish, the captured lookbehind
+						// belongs to partData
+						const _lookbehind = new Uint8Array(lookbehind.buffer, lookbehind.byteOffset, lookbehind.byteLength);
+						callback('partData', 0, previousIndex, _lookbehind);
+						previousIndex = 0;
+						mark('partData');
+
+						// reconsider the current character even so it interrupted the sequence
+						// it could be the beginning of a new sequence
+						i--;
+					}
+
+					break;
+				case S.END:
+					break;
+				default:
+					return i;
+			}
+		}
+
+		dataCallback('headerField');
+		dataCallback('headerValue');
+		dataCallback('partData');
+
+		this.index = index;
+		this.state = state;
+		this.flags = flags;
+
+		return length_;
+	}
+
+	end() {
+		function callback(self, name) {
+			const callbackSymbol = 'on' + name.slice(0, 1).toUpperCase() + name.slice(1);
+			if (callbackSymbol in self) {
+				self[callbackSymbol]();
+			}
+		}
+
+		if ((this.state === S.HEADER_FIELD_START && this.index === 0) ||
+			(this.state === S.PART_DATA && this.index === this.boundary.length)) {
+			callback(this, 'partEnd');
+			callback(this, 'end');
+		} else if (this.state !== S.END) {
+			return new Error('MultipartParser.end(): stream ended unexpectedly');
```

Comment:
The return value of `end()` is never checked, I think that this was meant to `throw`?

```suggestion
			throw new Error('MultipartParser.end(): stream ended unexpectedly');
```

Reply:
fixed